### DISCUSSION TITLE
Global loading indicator

### DISCUSSION THREAD

**Discussion by testerez** [vercel/swr#942] [Q&A]
Hello, I'd like to show a loading indicator whenever any swr request is in progress.

I came up with this solution:

```ts
const getIsSwrLoading = () => cache.keys().some(k => k.startsWith('validating@') && cache.get(k));

function useIsSwrLoading() {
  const [isLoading, setIsLoading] = useState(getIsSwrLoading());
  const update = () => setIsLoading(getIsSwrLoading());
  useEffect(() => cache.subscribe(update), []);
  return isLoading;
}
```
I have some questions about it:
- Does this approach look correct? Relying on undocumented cache keys sounds fragile.
- Would it make sense to have such a hook built in swr lib?
- Is cache documented anywhere

**Comment by testerez** [comment]
Actually, I just realized `react-query` has a very similar built-in hook: [useIsFetching](https://react-query.tanstack.com/guides/background-fetching-indicators#displaying-global-background-fetching-loading-state)

I like the drier API of swr, but I think this would be a useful addition.

**Reply by vikyw89** [reply]
any update on this ?

Having a global loading and or revalidating seems super useful and efficient

**Reply by NicoGrapsas** [reply]
same here, i think having useIsFecthing would be useful

**Comment by testerez** [comment]
I realize unmounted `useSWR` hooks can remain in `validating` state forever. So my solution is not actually suitable...

**Comment by iShawnWang** [comment]
I am also looking forward to this feature, in my opinion, it would be better to have a FSM, indicate : [initial, loading, error] state ~

**Comment by shuding** [comment]
It's a bit tricky to do this in the "React way", and some extra logic (reference counting) needs to be added to each SWR hook so it might bring some unneeded overhead for users that don't need it. 

I'll think about a better way, Middleware is probably doable.

**Reply by pke** [reply]
Since middleware are now possible in swr do you think it could be implemented with it?
To have a global "loading" bar/indicator would be nice to have in some apps.

**Comment by iShawnWang** [comment]
@pke since there is no official impl for now, we could try it ourself like this

**SAMPLE ONLY**

```
export default React.memo(() => {
  const [requestCount, setRequestCount] = useState(0);

  function requestCounter(useSWRNext) {
    return (key, fetcher, config) => {
      const extendedFetcher = (...args) => {
        setRequestCount((prev) => prev + 1);
        return fetcher(...args).finally(() =>
          setRequestCount((prev) => prev - 1)
        );
      };

      return useSWRNext(key, extendedFetcher, config);
    };
  }

  return (
    <SWRConfig value={{ use: [requestCounter] }}>
      <App />
      <Loading loading={requestCount > 0} />
    </SWRConfig>
  );
})

```

**Reply by pke** [reply]
Nice, thanks for the input! I have some fetchers that do not run network request but are like for async settings fetching or geolocation tracking. To not count them I guess I would have to put something in the key to hide them and detect them here in the middleware?

Maybe create a `useUntrackedSwr` with identical signature as `useSwr` that annotates the key (array) with a `"hidden"` string item.

**Reply by vikyw89** [reply]
i got this error, fetcher(...).finally is not a function

**Reply by vikyw89** [reply]
solved, it was caused by having a useSWR with undefined value as fetcher
